### ISSUE TITLE
incorrect escape in string? 

### ISSUE THREAD

**Issue by ajeecai** [eslint/espree#541]
The code is 
`ast = espree.parse(source, { comment: true, ecmaVersion: 'latest' });`

source is in [source.txt](https://github.com/eslint/espree/files/8262539/source.txt), the output ast is in [ast.txt](https://github.com/eslint/espree/files/8262550/ast.txt) . The value, compared to raw string,  looks not correct for escape, an example snapshot is: 

![image](https://user-images.githubusercontent.com/3023271/158602387-312479c5-11a1-4437-a45f-e98fd17b01e3.png)
 
Thanks

**Comment by mdjermanovic** [comment]
Can you provide a smaller example where this problem can be reproduced?

The part of the output you provided in the screenshot looks correct to me. For example, if this is the code being parsed:

```js
var foo = '<\/script>'
```

`\/` evaluates to `/` (it's a useless escape), so the literal's `value` is `</script>`, and `raw` is `'<\/script>'`. When represented in JSON or string literal notation, `\` needs to be escaped, so it would look like this:

```json
{
    "value": "</script>",
    "raw": "'<\\/script>'"
}
```

**Comment by ajeecai** [comment]
Hi @mdjermanovic , 

Yes I could give a simpler example as:

>       source = `var a = '<script src="//www.example.com"><\\/script>';`;
>       ast = espree.parse(source, { comment: true, ecmaVersion: 'latest' });

![image](https://user-images.githubusercontent.com/3023271/158734293-8086aa9f-6b3b-4448-a19b-05962f1c97b9.png)

To me, value is supposed to `"<script src=\"//www.example.com\"><\/script>"`

**Comment by mdjermanovic** [comment]
> To me, value is supposed to be `"<script src=\"//www.example.com\"><\/script>"`

It's the same value, because `"\/" === "/"`

```js
"<script src=\"//www.example.com\"><\/script>" === "<script src=\"//www.example.com\"></script>"; // true
```

`Literal.value` is the result of evaluating the literal. All literals that evaluate to the same value will have the same `Literal.value`. 

If you run this code:

```js
var a = '<\/script>';
console.log(a); // </script>
```

`'<\/script>'` is evaluated to a string value with 9 characters: `<`, `/`, `s`, `c`, `r`, `i`, `p`, `t`, `>`. There's no `\` because `\/` is unnecessarily escaped `/` and it evaluates to `/`.

**Comment by ajeecai** [comment]
Hi @mdjermanovic ,

For the first glance I admit what you said is correct for escape. But, consider `source `is containing string in a string ... let me give my use case for this weird problem:
```
  source = `var a = '<script src="//www.example.com"><\\/script>';`; <--- string in string
  ast = espree.parse(source, { comment: true, ecmaVersion: 'latest' });
```
Here is the server sending back to me: 
```
<!DOCTYPE html>
<html>
...
<script type="text/javascript">
    var a = '<script src="//www.example.com"><\/script>';
...
/script>
```
After calling `ast = espree.parse(source, { comment: true, ecmaVersion: 'latest' });`, we get the value and raw as figure in above comment.

After doing some stuff with `ast`,  calling `return this.stringifyNode(ast, true)` to serialize back, we get 
```
<!DOCTYPE html>
<html>
...
<script type="text/javascript">
    var a = '<script src="//www.example.com"></script>';
                                                 ^
                                                 |--- Different from original. 
                                                      When run in browser, there are hundreds of  script error
                                                      If we write it as <\/script>, it works perfectly.
...
/script>
```

Sorry if I am still not making you understand :)

Thanks

**Comment by mdjermanovic** [comment]
If `stringifyNode()` uses `Literal.value` then you'll probably need to post-process the output to avoid `</script>`.

> ```
> <!DOCTYPE html>
> <html>
> ...
> <script type="text/javascript">
>     var a = '<script src="//www.example.com"><\/script>';
> ...
> /script>
> ```
> 
> 
>     
>   
> 
> After calling `ast = espree.parse(source, { comment: true, ecmaVersion: 'latest' });`, we get the value and raw as figure in above comment:
> 
> 
> ```
> value: "<script src=\"//www.example.com\"></script>",
> raw: "'<script src=\"//www.example.com\"><\\/script>'",
> ```

This works correctly. `Literal.value` is a JavaScript value of the literal.

Consider the following:

```js
var a = '<script src="//www.example.com"></script>';

var a = '<script src="//www.example.com"><\/script>';

var a = '<script src="//www.example.com"><\x2Fscript>';

var a = '<script src="//www.example.com"><\u002Fscript>';
```

These are different string literals, but they evaluate to same JS string value because `"/"`, `"\/"`, `"\x2F"` and `"\u002F"` represent the same character, so they'll all have the same `Literal.value`. If you want to get the string literal representation from the original source code, you should use `Literal.raw`.

Here's an AST Explorer snippet with this example: https://astexplorer.net/#/gist/d763758a8fae9f3223a857e2d1d2d3dc/9827134a3236e25cfa2ed416498f396ca2f35415

**Comment by ajeecai** [comment]
Hi @mdjermanovic ,

I just simply wrap `stringifyNode` as
```
  private stringifyNode(node: estree.Node, comment?: boolean) {
    return escodegen.generate(node, { comment });
  }
```
where `escodegen` is from https://github.com/estools/escodegen.

But anyway, if you don't think it as a bug, please feel free to close it, I would tackle my problem with some workaround. 

Thank you

**Comment by mdjermanovic** [comment]
I'm closing this as Espree works as intended.

> I just simply wrap `stringifyNode` as
> 
> ```
>   private stringifyNode(node: estree.Node, comment?: boolean) {
>     return escodegen.generate(node, { comment });
>   }
> ```
> 
> where `escodegen` is from https://github.com/estools/escodegen.

You can also check in the [escodegen demo](https://estools.github.io/escodegen/demo/index.html) that it transforms `var a = '<\/script>';` to `var a = '</script>';`.

There is request for an option that would solve this problem https://github.com/estools/escodegen/issues/298, but since it's still open it looks like you'll have to manually preprocess or postprocess the code.
